### PR TITLE
feat(sdk-core): add specialized enable token functions

### PR DESCRIPTION
**Pull request by kevinwan-bitgo**
Ticket: BG-53196

### CODE REVIEW

File: modules/sdk-core/src/bitgo/wallet/iWallet.ts [hunk 1]
```diff
@@ -34,11 +34,16 @@ export interface BuildConsolidationTransactionOptions extends PrebuildTransactio
   consolidateAddresses?: string[];
 }
 
+export interface BuildTokenEnablementOptions extends PrebuildTransactionOptions {
+  tokens: string[];
```
Comment (joey-bai) [inline modules/sdk-core/src/bitgo/wallet/iWallet.ts:38]: `tokenNames`

File: modules/sdk-core/src/bitgo/wallet/wallet.ts [hunk 2]
```diff
@@ -2268,6 +2269,83 @@ export class Wallet implements IWallet {
     }
   }
 
+  public async buildTokenEnablements(
```
Comment (joey-bai) [inline modules/sdk-core/src/bitgo/wallet/wallet.ts:2272]: add JS docs for the functions you added

File: modules/sdk-core/src/bitgo/wallet/wallet.ts [hunk 3]
```diff
@@ -2268,6 +2269,83 @@ export class Wallet implements IWallet {
...
+    }
+    if (params.recipients) {
+      throw new Error('Can not specify recipients for token enablement transactions');
+    }
+
+    if (params.reqId) {
+      this.bitgo.setRequestTracer(params.reqId);
+    }
+
+    const buildParams = _.pick(params, this.prebuildWhitelistedParams());
```
Comment (joey-bai) [inline modules/sdk-core/src/bitgo/wallet/wallet.ts:2289]: do we want a separate set of whitelisted params just for token enablement?

File: modules/sdk-core/src/bitgo/wallet/wallet.ts [hunk 4]
```diff
@@ -2268,6 +2269,83 @@ export class Wallet implements IWallet {
...
+  }
+
+  public async sendTokenEnablement(params: PrebuildAndSignTransactionOptions = {}): Promise<any> {
+    if (!this.baseCoin.requiresTokenEnablement()) {
+      throw new Error(`${this.baseCoin.getFullName()} does not require token enablement transactions`);
+    }
+
+    if (typeof params.prebuildTx === 'string' || params.prebuildTx === undefined) {
+      throw new Error('Invalid build of token enablement.');
+    }
```
Comment (joey-bai) [inline modules/sdk-core/src/bitgo/wallet/wallet.ts:2310]: is there a way where we can check that the `prebuildTx` is a token enablement txn?

File: modules/sdk-core/src/bitgo/wallet/wallet.ts (hunk 3 above)
Comment (kevinwan-bitgo) [reply modules/sdk-core/src/bitgo/wallet/wallet.ts:2289]: `buildParams` here are the parameters that we plan on sending to `prebuildTransacation` so `prebuildWhitelistedParams` are the params that we want. We write token names into `recipients` later.

Comment (kevinwan-bitgo) [reply modules/sdk-core/src/bitgo/wallet/wallet.ts:2289]: Also, when we support the tss builds, we need to add the enable token params to `prebuildWhitelistedParams` for building with intent to work.

File: modules/sdk-core/src/bitgo/wallet/wallet.ts (hunk 4 above)
Comment (kevinwan-bitgo) [reply modules/sdk-core/src/bitgo/wallet/wallet.ts:2310]: `prebuildTransaction` does not return the `buildParams` used so this should be pretty difficult

Comment (kevinwan-bitgo) [reply modules/sdk-core/src/bitgo/wallet/wallet.ts:2310]: Another thing to note is that this function does not do anything specific to token enablements. It's the same as `sendAccountConsolidation` only we omit a few checks they do

Comment (joey-bai) [reply modules/sdk-core/src/bitgo/wallet/wallet.ts:2319]: 👍 thanks for explaining

**Review by joey-bai — APPROVED** [review]

File: modules/sdk-core/src/bitgo/baseCoin/iBaseCoin.ts [hunk 5]
```diff
@@ -321,6 +321,11 @@ export interface RecoverTokenTransaction {
   };
 }
 
+export interface TokenEnablementConfig {
+  requiresTokenEnablement: boolean;
+  supportsMultipleTokenEnablements?: boolean;
```
Comment (joey-bai) [inline modules/sdk-core/src/bitgo/baseCoin/iBaseCoin.ts:326]: add doc: "flag indicating if multiple tokens can be enabled in a single transaction"

File: modules/sdk-core/src/bitgo/baseCoin/iBaseCoin.ts [hunk 6]
```diff
@@ -321,6 +321,11 @@ export interface RecoverTokenTransaction {
   };
 }
 
+export interface TokenEnablementConfig {
+  requiresTokenEnablement: boolean;
```
Comment (joey-bai) [inline modules/sdk-core/src/bitgo/baseCoin/iBaseCoin.ts:325]: add doc: "flag indicating if tokens need to be enabled before receiving and sending"

File: modules/sdk-core/src/bitgo/baseCoin/iBaseCoin.ts (hunk 5 above)
Comment (joey-bai) [reply modules/sdk-core/src/bitgo/baseCoin/iBaseCoin.ts:326]: also, this should be required and we default to false.

File: modules/sdk-core/src/bitgo/wallet/iWallet.ts [hunk 7]
```diff
@@ -34,11 +34,19 @@ export interface BuildConsolidationTransactionOptions extends PrebuildTransactio
   consolidateAddresses?: string[];
 }
 
+export interface BuildTokenEnablementOptions extends PrebuildTransactionOptions {
+  tokens: {
+    name: string;
+    address?: string;
+  }[];
```
Comment (joey-bai) [inline modules/sdk-core/src/bitgo/wallet/iWallet.ts:41]: same lets document properties.

especially address as its not always clear. e.g. "Some coins like Solana require tokens be enabled on a specific address. If absent, we assume it is for the Wallet's root address"

**Review by joey-bai — APPROVED** [review]

**Review by mmcshinsky-bitgo — APPROVED** [review]

**Pull request merged**